src/com/amplitude/api/AmplitudeClient.java

```diff
@@ -703,6 +701,18 @@ public void run() {
             } else if (stringResponse.equals("request_db_write_failed")) {
                 Log.w(TAG,
                         "Couldn't write to request database on server, will attempt to reupload later");
+            } else if (response.code() == 413) {
+                // If blocked by one massive event, drop it
+                DatabaseHelper dbHelper = DatabaseHelper.getDatabaseHelper(context);
+                if (eventUploadMaxBatchSize == 1) {
+                    dbHelper.removeEvents(maxId);
+                }
+
+                // Server complained about length of request, backoff and try again later
+                int numEvents = Math.min((int)dbHelper.getEventCount(), eventUploadMaxBatchSize);
+                eventUploadMaxBatchSize = (int)Math.ceil(numEvents / 2.0);
```

Comment:
so eventUploadMaxBatchSize persists until the application falls out of memory, which we don't really have control over. Thinking more I think we want the post request to fully resolve all events up to maxId, and have the request size reduction temporary to a single call of makeEventUploadPostRequest